src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -22127,6 +22127,1449 @@ void Main(string? s)
+                // (26,15): warning CS8604: Possible null reference argument for parameter 't' in 'T Program.F2<T>(T t)'.
+                //         F2<T>(t3).ToString(); // 6
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "t3").WithArguments("t", "T Program.F2<T>(T t)").WithLocation(26, 15),
+                // (27,15): warning CS8604: Possible null reference argument for parameter 't' in 'T? Program.F3<T>(T t)'.
+                //         F3<T>(t3).ToString(); // 7
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "t3").WithArguments("t", "T? Program.F3<T>(T t)").WithLocation(27, 15));
+        }
+
+        [Fact]
+        public void NotNull_ReturnValue_03()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    [return: NotNull] static string F1() => string.Empty;
+    [return: NotNull] static string? F2() => string.Empty;
+    static void M()
+    {
+        F1().ToString();
+        F2().ToString();
+    }
+}";
+            var comp = CreateCompilation(new[] { NotNullAttributeDefinition, source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NotNull_ReturnValue_04()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    [return: NotNull] static int F1() => 1;
+    [return: NotNull] static int? F2() => 2;
+    static void M()
+    {
+        F1().ToString();
+        _ = F2().Value;
+    }
+}";
+            var comp = CreateCompilation(new[] { NotNullAttributeDefinition, source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NotNull_ReturnValue_05()
+        {
+            var source0 =
+@"using System.Runtime.CompilerServices;
+public class A
+{
+    public static T F1<T>(T t) where T : class => t;
+    [return: NotNull] public static T F2<T>(T t) where T : class => t;
+}";
+            var comp = CreateCompilation(new[] { NotNullAttributeDefinition, source0 }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+            var ref0 = comp.EmitToImageReference();
+
+            var source =
+@"class B : A
+{
+    static void Main()
+    {
+        object x = null; // 1
+        object? y = new object();
+        F1(x).ToString(); // 2
+        F1(y).ToString();
+        F2(x).ToString();
+        F2(y).ToString();
+    }
+}";
+            comp = CreateCompilation(source, references: new[] { ref0 }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (5,20): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         object x = null; // 1
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(5, 20),
+                // (7,9): warning CS
```

Comment:
>y1.ToString(); [](start = 27, length = 14)

:memo: I'm investigating this. I'll push a commit to fix.

Reply:
I believe this is correct.  The argument type and state is `object?`, so that is what we infer for `T`.  Then we report the constraint failure.  Since we don't use constraints in type inference, I believe this is correct.

---
In reply to: [288805395](https://github.com/dotnet/roslyn/pull/35955#discussion_r288805395) [](ancestors = 288805395)